src/node_crypto.cc

```diff
@@ -1631,6 +1660,76 @@ void SSLWrap<Base>::IsInitFinished(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+int compar(const void* a, const void* b) {
+  return memcmp(a, b, CNNIC_WHITELIST_HASH_LEN);
+}
+
+
+int IsSelfSigned(X509* cert) {
+  return X509_NAME_cmp(X509_get_subject_name(cert),
+                       X509_get_issuer_name(cert)) == 0;
+}
+
+
+X509* LookforRoot(STACK_OF(X509)* sk) {
+  int i;
+  X509* cert;
+  for(i = 0; i < sk_X509_num(sk); i++) {
+    cert = sk_X509_value(sk, i);
+    if (IsSelfSigned(cert))
+      return cert;
+  }
+  return nullptr;
+}
+
+
+// Whitelist check for certs issued by CNNIC. See
+// https://blog.mozilla.org/security/2015/04/02/distrusting-new-cnnic-certificates/
+bool CheckWhiteList(SSL* ssl) {
+  X509* root_cert;
+  X509_NAME* root_name;
+  unsigned char hash[CNNIC_WHITELIST_HASH_LEN];
+  unsigned int hashlen = CNNIC_WHITELIST_HASH_LEN;
+
+  if (SSL_is_server(ssl))
+    return true;
+
+  STACK_OF(X509)* chain = SSL_get_peer_cert_chain(ssl);
+
+  // take the last cert as root at the first time
+  root_cert = sk_X509_value(chain, sk_X509_num(chain)-1);
+  root_name = X509_get_subject_name(root_cert);
+
+  if (!IsSelfSigned(root_cert)) {
+    root_cert = LookforRoot(chain);
+    if(root_cert == nullptr)
+      return true;
+
+    root_name = X509_get_subject_name(root_cert);
+  }
+
+  if (X509_NAME_cmp(root_name, cnnic_name) == 0 ||
+      X509_NAME_cmp(root_name, cnnic_ev_name) == 0) {
+
+    // When the cert is issued from either CNNNIC ROOT CA or CNNNIC EV
+    // ROOT CA, check a hash of its leaf cert if it is in the
+    // whitelist.
+    int ret = X509_digest(sk_X509_value(chain, 0), EVP_sha256(), hash, &hashlen);
```

Comment:
This is the same as Chrome does in https://code.google.com/p/chromium/codesearch#chromium/src/net/cert/x509_certificate_openssl.cc&l=236